server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesSourceConfig.java

```diff
@@ -19,13 +19,29 @@
 
 package org.elasticsearch.search.aggregations.bucket.composite;
 
+import org.apache.lucene.index.IndexReader;
 import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.util.BigArrays;
 import org.elasticsearch.index.mapper.MappedFieldType;
 import org.elasticsearch.search.DocValueFormat;
 import org.elasticsearch.search.aggregations.support.ValuesSource;
 import org.elasticsearch.search.sort.SortOrder;
 
-class CompositeValuesSourceConfig {
+import java.util.function.LongConsumer;
+
+public class CompositeValuesSourceConfig {
+
+    @FunctionalInterface
+    public interface SingleDimensionValuesSourceProvider {
+        SingleDimensionValuesSource<?> createValuesSource(
+            BigArrays bigArrays,
+            IndexReader reader,
+            int size,
+            LongConsumer addRequestCircuitBreakerBytes,
+            CompositeValuesSourceConfig config
```

Comment:
It's definitely a little weird that `CompositeValuesSourceConfig` is both the object that holds this function, and a parameter to this function.  I'm pretty confidant that there's a less tangled way to do this.  My plan is to land this initial change to get stuff into the registry at all, and try to neaten up this edge in a follow up PR.